### PR TITLE
#8591: yolox_m reference and ttnn implementation

### PR DESCRIPTION
**Pull request by saichandax**
It has reference and ttnn implementation for the yolox_m implementation.

Perf results

Focus submodule: [ops_perf_results_Focus_2024_06_24_09_47_09.csv](https://github.com/user-attachments/files/15953610/ops_perf_results_Focus_2024_06_24_09_47_09.csv)

Dark2 submodule: [ops_perf_results_Dark2_2024_06_24_09_48_01.csv](https://github.com/user-attachments/files/15953623/ops_perf_results_Dark2_2024_06_24_09_48_01.csv)

Dark3 submodule: [ops_perf_results_Dark3_2024_06_24_09_49_20.csv](https://github.com/user-attachments/files/15953746/ops_perf_results_Dark3_2024_06_24_09_49_20.csv)

Dark4 submodule: [ops_perf_results_Dark4_2024_06_24_09_53_37.csv](https://github.com/user-attachments/files/15953752/ops_perf_results_Dark4_2024_06_24_09_53_37.csv)

Dark5 submodule: [ops_perf_results_Dark5_2024_06_24_09_54_19.csv](https://github.com/user-attachments/files/15953759/ops_perf_results_Dark5_2024_06_24_09_54_19.csv)

YOLOPAFPN module: [ops_perf_results_YoloPAFPN_2024_06_24_10_08_04.csv](https://github.com/user-attachments/files/15953762/ops_perf_results_YoloPAFPN_2024_06_24_10_08_04.csv)

Head submodule: [ops_perf_results_Head_2024_06_24_09_57_34.csv](https://github.com/user-attachments/files/15953768/ops_perf_results_Head_2024_06_24_09_57_34.csv)

Yolox model: [ops_perf_results_Yolox_2024_06_24_09_34_18.csv](https://github.com/user-attachments/files/15953776/ops_perf_results_Yolox_2024_06_24_09_34_18.csv)

Note:
In demo, real image input pcc
PCC:
Out1: 0.9117486583752237
Out2: 0.932798661586333
Out3: 0.9420547476555219

Demo results: [here](https://docs.google.com/document/d/1WntZDmsJr_KPUY89De0D1GvwR8m7u7MzCHtrz6TvZLs/edit)

### CODE REVIEW

**Review by saichandax — COMMENTED** [review]
rename this file to `cpu_demo.py`

File: models/experimental/functional_yolox_m/demo/demo.py [hunk 1]
```diff
@@ -0,0 +1,593 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+
+# SPDX-License-Identifier: Apache-2.0
+
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/demo/demo.py:4]: Rename the file to `cpu_demo.py`

File: models/experimental/functional_yolox_m/demo/demo.py [hunk 2]
```diff
@@ -0,0 +1,593 @@
...
+from models.experimental.functional_yolox_m.tt.ttnn_cspdarknet import TtCSPDarknet
+
+import tests.ttnn.integration_tests.yolox_m.custom_preprocessor_d2 as D2
+import tests.ttnn.integration_tests.yolox_m.custom_preprocessor_d3 as D3
+import tests.ttnn.integration_tests.yolox_m.custom_preprocessor_d4 as D4
+import tests.ttnn.integration_tests.yolox_m.custom_preprocessor_d5 as D5
+import tests.ttnn.integration_tests.yolox_m.custom_preprocessor_cspdarknet as cspdarknet
+import tests.ttnn.integration_tests.yolox_m.custom_preprocessor_yolopafpn as fpn
+import tests.ttnn.integration_tests.yolox_m.custom_preprocessor_yolohead as head
+import ttnn
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/demo/demo.py:32]: Are we using all the imports?

File: models/experimental/functional_yolox_m/demo/demo.py [hunk 3]
```diff
@@ -0,0 +1,593 @@
...
+        cv2.rectangle(img, (x0, y0), (x1, y1), color, 2)
+
+        txt_bk_color = (_COLORS[cls_id] * 255 * 0.7).astype(np.uint8).tolist()
+        cv2.rectangle(img, (x0, y0 + 1), (x0 + txt_size[0] + 1, y0 + int(1.5 * txt_size[1])), txt_bk_color, -1)
+        cv2.putText(img, text, (x0, y0 + txt_size[1]), font, 0.4, txt_color, thickness=1)
+
+    return img
+
+
+_COLORS = (
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/demo/demo.py:145]: Can we load class names and colors from a different file?

File: models/experimental/functional_yolox_m/demo/demo.py [hunk 4]
```diff
@@ -0,0 +1,593 @@
...
+            0.5,
+            0,
+        ]
+    )
+    .astype(np.float32)
+    .reshape(-1, 3)
+)
+
+
+def preproc(img, input_size, swap=(2, 0, 1)):
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/demo/demo.py:395]: can we have method name as `def preprocess(...)`?

File: models/experimental/functional_yolox_m/demo/demo.py [hunk 5]
```diff
@@ -0,0 +1,593 @@
...
+    if decode_in_inference:
+        outputs = decode_outputs(outputs, dtype=img.type(), hw=hw)
+    outputs = postprocess(outputs, num_classes, confthre, nmsthre, class_agnostic=True)
+
+    result_image = visual(outputs[0], img_info, confthre)
+    save_file_name = "./torch_dog.jpg"
+    cv2.imwrite(save_file_name, result_image)
+
+
+import pytest
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/demo/demo.py:568]: keep imports at one place.

File: models/experimental/functional_yolox_m/demo/demo.py [hunk 6]
```diff
@@ -0,0 +1,593 @@
...
+
+
+# @pytest.mark.parametrize("device_l1_small_size", [32768], indirect=True)
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/demo/demo.py:571]: remove if not used

File: models/experimental/functional_yolox_m/demo/demo_ttnn.py [hunk 7]
```diff
@@ -0,0 +1,677 @@
...
+from models.experimental.functional_yolox_m.tt.ttnn_cspdarknet import TtCSPDarknet
+
+import tests.ttnn.integration_tests.yolox_m.custom_preprocessor_d2 as D2
+import tests.ttnn.integration_tests.yolox_m.custom_preprocessor_d3 as D3
+import tests.ttnn.integration_tests.yolox_m.custom_preprocessor_d4 as D4
+import tests.ttnn.integration_tests.yolox_m.custom_preprocessor_d5 as D5
+import tests.ttnn.integration_tests.yolox_m.custom_preprocessor_cspdarknet as cspdarknet
+import tests.ttnn.integration_tests.yolox_m.custom_preprocessor_yolopafpn as fpn
+import tests.ttnn.integration_tests.yolox_m.custom_preprocessor_yolohead as head
+import ttnn
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/demo/demo_ttnn.py:33]: Are we using all the imports? Remove unnecessary imports

File: models/experimental/functional_yolox_m/demo/demo_ttnn.py [hunk 8]
```diff
@@ -0,0 +1,677 @@
...
+        cv2.rectangle(img, (x0, y0), (x1, y1), color, 2)
+
+        txt_bk_color = (_COLORS[cls_id] * 255 * 0.7).astype(np.uint8).tolist()
+        cv2.rectangle(img, (x0, y0 + 1), (x0 + txt_size[0] + 1, y0 + int(1.5 * txt_size[1])), txt_bk_color, -1)
+        cv2.putText(img, text, (x0, y0 + txt_size[1]), font, 0.4, txt_color, thickness=1)
+
+    return img
+
+
+_COLORS = (
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/demo/demo_ttnn.py:146]: Can we load class names and colors from a different file?

File: models/experimental/functional_yolox_m/demo/demo_ttnn.py [hunk 9]
```diff
@@ -0,0 +1,677 @@
...
+    "clock",
+    "vase",
+    "scissors",
+    "teddy bear",
+    "hair drier",
+    "toothbrush",
+)
+
+
+def vis(img, boxes, scores, cls_ids, conf=0.5, class_names=None):
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/demo/demo_ttnn.py:119]: Can we have a file `demo_utils.py` and keep all the utility methods in that file?

File: models/experimental/functional_yolox_m/demo/demo_ttnn.py [hunk 10]
```diff
@@ -0,0 +1,677 @@
...
+            0.5,
+            0,
+        ]
+    )
+    .astype(np.float32)
+    .reshape(-1, 3)
+)
+
+
+def preproc(img, input_size, swap=(2, 0, 1)):
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/demo/demo_ttnn.py:396]: `def preprocess(...)`?

File: models/experimental/functional_yolox_m/demo/demo_ttnn.py [hunk 11]
```diff
@@ -0,0 +1,677 @@
...
+    bboxes /= ratio
+
+    cls = output[:, 6]
+    scores = output[:, 4] * output[:, 5]
+
+    vis_res = vis(img, bboxes, scores, cls, cls_conf, cls_names)
+    return vis_res
+
+
+def postprocess(prediction, num_classes, conf_thre=0.7, nms_thre=0.45, class_agnostic=False):
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/demo/demo_ttnn.py:467]: There is a lot of repitition of code across demos. Please move helper methods to demo_utils file.

File: models/experimental/functional_yolox_m/demo/demo_ttnn.py [hunk 12]
```diff
@@ -0,0 +1,677 @@
...
+                device, model.backbone.backbone, name, ttnn_module_args["backbone"]["backbone"]
+            )
+            parameters["backbone"].update(
+                fpn.custom_preprocessor(device, model.backbone, name, ttnn_module_args["backbone"])
+            )
+            parameters["head"] = head.custom_preprocessor(device, model.head, name, ttnn_module_args["head"])
+        return parameters
+
+    return custom_preprocessor
+
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/demo/demo_ttnn.py:636]: remove new line

File: models/experimental/functional_yolox_m/demo/demo_ttnn.py [hunk 13]
```diff
@@ -0,0 +1,677 @@
...
+        initialize_model=lambda: torch_model,
+        run_model=lambda model: model(torch_input_tensor),
+        custom_preprocessor=create_custom_preprocessor(device),
+        reader_patterns_cache=reader_patterns_cache,
+        device=device,
+    )
+    ttnn_model = TtYOLOX(device, parameters)
+    inference(ttnn_model, device)
+
+    # assert_with_pcc(torch_output_tensor0, output_tensor0, pcc=0.99)  # PCC = 0.8041133430120856
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/demo/demo_ttnn.py:675]: Remove comments if not needed

File: models/experimental/functional_yolox_m/reference/bottleneck_block.py [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+        for i in range(nblocks):
+            conv1 = nn.Conv2d(ch, ch, 1, 1, 0, bias=False)
+            bn1 = nn.BatchNorm2d(ch, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
+            relu = nn.SiLU(inplace=True)
+            conv2 = nn.Conv2d(ch, ch, 3, 1, 1, bias=False)
+            bn2 = nn.BatchNorm2d(ch, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
+            resblock_one = nn.ModuleList([conv1, bn1, relu, conv2, bn2, relu])
+            self.module_list.append(resblock_one)
+
+    def forward(self, x):
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/reference/bottleneck_block.py:23]: Add type hinting for all the forward or __call__ methods:

`def forward( self, x: torch.Tensor): . . .`

File: models/experimental/functional_yolox_m/reference/csp_darknet.py [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+
+import torch.nn as nn
+from models.experimental.functional_yolox_m.reference.bottleneck_block import BottleNeckBlock
+from models.experimental.functional_yolox_m.reference.dark2 import Focus, Dark2
+from models.experimental.functional_yolox_m.reference.dark3 import Dark3
+from models.experimental.functional_yolox_m.reference.dark4 import Dark4
+from models.experimental.functional_yolox_m.reference.dark5 import Dark5
+from models.utility_functions import comp_pcc
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/reference/csp_darknet.py:12]: remove this import.

File: models/experimental/functional_yolox_m/reference/csp_darknet.py [hunk 16]
```diff
@@ -0,0 +1,39 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+
+# SPDX-License-Identifier: Apache-2.0
+
+
+import torch.nn as nn
+from models.experimental.functional_yolox_m.reference.bottleneck_block import BottleNeckBlock
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/reference/csp_darknet.py:7]: Is bottleneck_block used in this file?

File: models/experimental/functional_yolox_m/reference/csp_darknet.py [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+    def __init__(self):
+        super(CSPDarknet, self).__init__()
+
+        self.stem = Focus()
+        self.dark2 = Dark2()
+        self.dark3 = Dark3()
+        self.dark4 = Dark4()
+        self.dark5 = Dark5()
+
+    def forward(self, x):
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/reference/csp_darknet.py:25]: Add type hinting

File: models/experimental/functional_yolox_m/reference/dark3.py [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+        self.b3 = nn.BatchNorm2d(96, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
+
+        self.c4 = nn.Conv2d(192, 192, kernel_size=1, stride=1, bias=False)
+        self.b4 = nn.BatchNorm2d(192, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
+
+        self.bblock = BottleNeckBlock(96, 6, True)
+        # Activation function
+        self.silu = nn.SiLU(inplace=True)
+
+    def forward(self, x):
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/reference/dark3.py:33]: Add type hinting here

File: models/experimental/functional_yolox_m/reference/dark4.py [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+
+        self.c4 = nn.Conv2d(384, 384, kernel_size=1, stride=1, bias=False)
+        self.b4 = nn.BatchNorm2d(384, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
+
+        self.bblock = BottleNeckBlock(192, 6, True)
+
+        # Activation function
+        self.silu = nn.SiLU(inplace=True)
+
+    def forward(self, x):
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/reference/dark4.py:34]: Add type hinting here

File: models/experimental/functional_yolox_m/reference/dark2.py [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+        self.b3 = nn.BatchNorm2d(48, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
+
+        self.c4 = nn.Conv2d(96, 96, kernel_size=1, stride=1, bias=False)
+        self.b4 = nn.BatchNorm2d(96, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
+
+        self.bblock = BottleNeckBlock(48, 2, True)
+        # Activation function
+        self.silu = nn.SiLU(inplace=True)
+
+    def forward(self, x):
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/reference/dark2.py:62]: Add type hinting here

File: models/experimental/functional_yolox_m/reference/dark5.py [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+        self.b5 = nn.BatchNorm2d(384, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
+
+        self.c6 = nn.Conv2d(768, 768, kernel_size=1, stride=1, bias=False)
+        self.b6 = nn.BatchNorm2d(768, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
+
+        self.bblock = BottleNeckBlock(384, 2, False)
+        # Activation function
+        self.silu = nn.SiLU(inplace=True)
+
+    def forward(self, x):
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/reference/dark5.py:43]: Add type hinting here

File: models/experimental/functional_yolox_m/reference/dark5.py [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+        bblock_out = self.bblock(x4_m)
+
+        conc2 = torch.cat((bblock_out, x5_m), dim=1)
+        x6 = self.c6(conc2)
+        x6_b = self.b6(x6)
+        x6_m = self.silu(x6_b)
+        import ttnn
+
+        input_tensor = ttnn.from_torch(x3_m, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT)
+        ttnn.dump_tensor("tests/ttnn/integration_tests/yolox_m/dark5_conv4_inp_torch.pt", input_tensor)
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/reference/dark5.py:81]: Is it required in the pipeline?

File: models/experimental/functional_yolox_m/reference/yolo_head.py [hunk 23]
```diff
@@ -0,0 +1,191 @@
...
+        self.c22 = nn.Conv2d(192, 192, kernel_size=1, stride=1, bias=False)
+        self.b22 = nn.BatchNorm2d(192, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
+
+        self.c23 = nn.Conv2d(384, 192, kernel_size=1, stride=1, bias=False)
+        self.b23 = nn.BatchNorm2d(192, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
+
+        self.c24 = nn.Conv2d(768, 192, kernel_size=1, stride=1, bias=False)
+        self.b24 = nn.BatchNorm2d(192, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
+
+    def forward(self, x):
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/reference/yolo_head.py:80]: Add type hinting here

File: models/experimental/functional_yolox_m/reference/yolo_pafpn.py [hunk 24]
```diff
@@ -0,0 +1,191 @@
...
+        self.b16 = nn.BatchNorm2d(768, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
+
+        self.bblock4 = BottleNeckBlock(384, 2, False)
+
+        # Activation function
+        self.silu = nn.SiLU(inplace=True)
+
+        self.in_features = ("dark3", "dark4", "dark5")
+
+    def forward(self, input):
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/reference/yolo_pafpn.py:87]: Add type hinting here

File: models/experimental/functional_yolox_m/reference/yolox_m.py [hunk 25]
```diff
@@ -0,0 +1,24 @@
...
+from models.experimental.functional_yolox_m.reference.yolo_head import YOLOXHead
+
+
+class YOLOX(nn.Module):
+    def __init__(self, backbone=None, head=None):
+        super().__init__()
+        self.backbone = YOLOPAFPN()
+        self.head = YOLOXHead(80)
+
+    def forward(self, x, targets=None):
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/reference/yolox_m.py:18]: Add type hinting here

File: models/experimental/functional_yolox_m/tt/ttnn_bottleneck_block.py [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+        self.nblocks = nblocks
+        self.module_list = []
+        keys = list(parameters.keys())
+        for i in range(0, len(keys), 2):
+            conv1 = parameters[keys[i]]
+            conv2 = parameters[keys[i + 1]]
+            resblock_one = [conv1, conv2]
+            self.module_list.append(resblock_one)
+
+    def __call__(self, device, input_tensor):
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/tt/ttnn_bottleneck_block.py:22]: Add type hinting here:
`def __call__(self, device, input_tensor: ttnn.Tensor)`

File: models/experimental/functional_yolox_m/tt/ttnn_cspdarknet.py [hunk 27]
```diff
@@ -0,0 +1,40 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+
+# SPDX-License-Identifier: Apache-2.0
+
+
+import ttnn
+import tt_lib
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/tt/ttnn_cspdarknet.py:7]: Is this import needed?

File: models/experimental/functional_yolox_m/tt/ttnn_cspdarknet.py [hunk 28]
```diff
@@ -0,0 +1,40 @@
...
+        device,
+        parameters,
+    ) -> None:
+        self.focus = TtFocus(parameters["stem"])
+        self.dark2 = TtDark2(parameters["dark2"])
+        self.dark3 = TtDark3(parameters["dark3"])
+        self.dark4 = TtDark4(parameters["dark4"])
+        self.dark5 = TtDark5(device, parameters["dark5"])
+
+    def __call__(self, device, input_tensor):
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/tt/ttnn_cspdarknet.py:26]: Add type hinting here

File: models/experimental/functional_yolox_m/tt/ttnn_dark2.py [hunk 29]
```diff
@@ -0,0 +1,123 @@
...
+
+
+class TtFocus:
+    def __init__(
+        self,
+        parameters,
+    ) -> None:
+        self.c1 = parameters.c1
+
+    def __call__(self, device, input_tensor):
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/tt/ttnn_dark2.py:19]: In all the __call__ methods, we need to add type hinting. Please check.

File: tests/ttnn/integration_tests/yolox_m/README.md [hunk 30]
```diff
@@ -0,0 +1,10 @@
+# Yolox_m ttnn implementation
+
+## Current status
+
+1. ttnn.conv in focus sub-module has PCC of 0.99. However, torch slice is used in this sub-module, because reference has this type of slice x[..., ::2, ::2] where ::2 indicates alternative elements will be chosen, and its throwing error in ::2.
```
Comment (saichandax) [inline tests/ttnn/integration_tests/yolox_m/README.md:5]: This README is out-dated. Please update with proper instructions on how to run the tests.
Also add a README for the demo separately. Refer to existing Demo README for any other model for reference.

File: tests/ttnn/integration_tests/yolox_m/custom_preprocessor_d4.py [hunk 31]
```diff
@@ -0,0 +1,94 @@
...
+            ttnn_module_args["bblock"][f"bblock_{i}_conv2"]["weights_dtype"] = ttnn.bfloat8_b
+            ttnn_module_args["bblock"][f"bblock_{i}_conv2"]["use_shallow_conv_variant"] = False
+            weight2, bias2 = fold_batch_norm2d_into_conv2d(conv2, bn2)
+            update_ttnn_module_args(ttnn_module_args["bblock"][f"bblock_{i}_conv2"])
+            parameters["bblock"][f"bblock_{i}_conv2"], _ = preprocess_conv2d(
+                weight2, bias2, ttnn_module_args["bblock"][f"bblock_{i}_conv2"], return_parallel_config=True
+            )
+
+        ttnn_module_args.c4["use_shallow_conv_variant"] = False
+        # ttnn_module_args.c4["use_shallow_conv_variant"] = (
```
Comment (saichandax) [inline tests/ttnn/integration_tests/yolox_m/custom_preprocessor_d4.py:84]: Remove unnecessary commetns

File: tests/ttnn/integration_tests/yolox_m/custom_preprocessor_yolohead.py [hunk 32]
```diff
@@ -0,0 +1,230 @@
...
+        ttnn_module_args.c13["weights_dtype"] = ttnn.bfloat8_b
+        # conv13_weight, conv13_bias = fold_batch_norm2d_into_conv2d(model.c13, model.b13)
+        update_ttnn_module_args(ttnn_module_args.c13)
+        parameters["c13"], c13_parallel_config = preprocess_conv2d(
+            model.c13.weight, model.c13.bias, ttnn_module_args.c13, return_parallel_config=True
+        )
+
+        ttnn_module_args.c14["use_shallow_conv_variant"] = False
+        ttnn_module_args.c14["weights_dtype"] = ttnn.bfloat8_b
+        # conv14_weight, conv14_bias = fold_batch_norm2d_into_conv2d(model.c14, model.b14)
```
Comment (saichandax) [inline tests/ttnn/integration_tests/yolox_m/custom_preprocessor_yolohead.py:144]: Remove unnecessary comments.

File: tests/ttnn/integration_tests/yolox_m/custom_preprocessor_yolopafpn.py [hunk 33]
```diff
@@ -0,0 +1,286 @@
...
+        #     False if device.arch() == tt_lib.device.Arch.WORMHOLE_B0 else True
+        # )
+        ttnn_module_args.c3["use_shallow_conv_variant"] = False
+        conv3_weight, conv3_bias = fold_batch_norm2d_into_conv2d(model.c3, model.b3)
+        update_ttnn_module_args(ttnn_module_args.c3)
+        parameters["c3"], c3_parallel_config = preprocess_conv2d(
+            conv3_weight, conv3_bias, ttnn_module_args.c3, return_parallel_config=True
+        )
+
+        # ttnn_module_args.c4["use_shallow_conv_variant"] = (
```
Comment (saichandax) [inline tests/ttnn/integration_tests/yolox_m/custom_preprocessor_yolopafpn.py:57]: Remove unnecessary comments.

File: tests/ttnn/integration_tests/yolox_m/test_ttnn_dark2.py [hunk 34]
```diff
@@ -0,0 +1,87 @@
...
+def create_custom_preprocessor(device):
+    def custom_preprocessor(model, name, ttnn_module_args):
+        parameters = {}
+        parameters = D2.custom_preprocessor(device, model, name, ttnn_module_args)
+        return parameters
+
+    return custom_preprocessor
+
+
+import pytest
```
Comment (saichandax) [inline tests/ttnn/integration_tests/yolox_m/test_ttnn_dark2.py:30]: Keep imports at one place and check if we are using all the imports. Remove any unnecessary imports.

File: tests/ttnn/integration_tests/yolox_m/test_ttnn_dark2.py [hunk 35]
```diff
@@ -0,0 +1,87 @@
...
+    output_tensor = ttnn_model(device, input_tensor)
+
+    #
+    # Tensor Postprocessing
+    #
+    output_tensor = ttnn.to_torch(output_tensor)
+    output_tensor = output_tensor.reshape(1, 160, 160, 96)
+    output_tensor = torch.permute(output_tensor, (0, 3, 1, 2))
+    output_tensor = output_tensor.to(torch_input_tensor.dtype)
+    assert_with_pcc(torch_output_tensor, output_tensor, pcc=0.99)  # PCC = 0.9852007334820249
```
Comment (saichandax) [inline tests/ttnn/integration_tests/yolox_m/test_ttnn_dark2.py:87]: If PCC is ~0.98, test might fail as we are setting 0.99
Set appropriate passing PCC as threshold.

File: tests/ttnn/integration_tests/yolox_m/test_ttnn_dark3.py [hunk 36]
```diff
@@ -0,0 +1,88 @@
...
+    output_tensor = ttnn_model(device, input_tensor)
+
+    #
+    # Tensor Postprocessing
+    #
+    output_tensor = ttnn.to_torch(output_tensor)
+    output_tensor = output_tensor.reshape(1, 80, 80, 192)
+    output_tensor = torch.permute(output_tensor, (0, 3, 1, 2))
+    output_tensor = output_tensor.to(torch_input_tensor.dtype)
+    assert_with_pcc(torch_output_tensor, output_tensor, pcc=0.99)  # PCC = 0.9804122271033656
```
Comment (saichandax) [inline tests/ttnn/integration_tests/yolox_m/test_ttnn_dark3.py:88]: Same here. set passing PCC

File: tests/ttnn/integration_tests/yolox_m/test_ttnn_dark4.py [hunk 37]
```diff
@@ -0,0 +1,88 @@
...
+    output_tensor = ttnn_model(device, input_tensor)
+
+    #
+    # Tensor Postprocessing
+    #
+    output_tensor = ttnn.to_torch(output_tensor)
+    output_tensor = output_tensor.reshape(1, 40, 40, 384)
+    output_tensor = torch.permute(output_tensor, (0, 3, 1, 2))
+    output_tensor = output_tensor.to(torch_input_tensor.dtype)
+    assert_with_pcc(torch_output_tensor, output_tensor, pcc=0.99)  # PCC = 0.9824520887645896
```
Comment (saichandax) [inline tests/ttnn/integration_tests/yolox_m/test_ttnn_dark4.py:88]: Same here.

File: tests/ttnn/integration_tests/yolox_m/test_ttnn_dark5.py [hunk 38]
```diff
@@ -0,0 +1,88 @@
...
+    output_tensor = ttnn_model(device, input_tensor)
+
+    #
+    # Tensor Postprocessing
+    #
+    output_tensor = ttnn.to_torch(output_tensor)
+    output_tensor = output_tensor.reshape(1, 20, 20, 768)
+    output_tensor = torch.permute(output_tensor, (0, 3, 1, 2))
+    output_tensor = output_tensor.to(torch_input_tensor.dtype)
+    assert_with_pcc(torch_output_tensor, output_tensor, pcc=0.99)  # PCC = 0.9564151119600169
```
Comment (saichandax) [inline tests/ttnn/integration_tests/yolox_m/test_ttnn_dark5.py:88]: Same here

File: tests/ttnn/integration_tests/yolox_m/test_ttnn_yolopafpn.py [hunk 39]
```diff
@@ -0,0 +1,144 @@
...
+    output_tensor1 = output_tensor1.to(torch_input_tensor.dtype)
+
+    output_tensor2 = ttnn.to_torch(output_tensor2)
+    output_tensor2 = output_tensor2.reshape(1, 20, 20, 768)
+    output_tensor2 = torch.permute(output_tensor2, (0, 3, 1, 2))
+    output_tensor2 = output_tensor2.to(torch_input_tensor.dtype)
+
+    assert_with_pcc(torch_output_tensor0, output_tensor0, pcc=0.99)  # PCC = 0.5187837901243553
+    assert_with_pcc(torch_output_tensor1, output_tensor1, pcc=0.99)  # PCC = 0.4114014847628962
+    assert_with_pcc(torch_output_tensor2, output_tensor2, pcc=0.99)  # PCC = 0.4744938004335893
```
Comment (saichandax) [inline tests/ttnn/integration_tests/yolox_m/test_ttnn_yolopafpn.py:144]: Set passing PCC. If comment is not applicable, remove it.

File: tests/ttnn/integration_tests/yolox_m/test_ttnn_yolox.py [hunk 40]
```diff
@@ -0,0 +1,152 @@
...
+    output_tensor1 = output_tensor1.to(torch_input_tensor.dtype)
+
+    output_tensor2 = ttnn.to_torch(output_tensor2)
+    output_tensor2 = output_tensor2.reshape(1, 20, 20, 85)
+    output_tensor2 = torch.permute(output_tensor2, (0, 3, 1, 2))
+    output_tensor2 = output_tensor2.to(torch_input_tensor.dtype)
+
+    assert_with_pcc(torch_output_tensor0, output_tensor0, pcc=0.99)  # PCC = 0.8041133430120856
+    assert_with_pcc(torch_output_tensor1, output_tensor1, pcc=0.99)  # PCC = 0.8681695629677733
+    assert_with_pcc(torch_output_tensor2, output_tensor2, pcc=0.99)  # PCC = 0.9175503378141336
```
Comment (saichandax) [inline tests/ttnn/integration_tests/yolox_m/test_ttnn_yolox.py:152]: Remove comments if not applicable. Set passing PCC for threshold.

File: tests/ttnn/integration_tests/yolox_m/test_yolox_flops.py [hunk 41]
```diff
@@ -0,0 +1,17 @@
...
+
+torch_model = YOLOX()
+ank
+torch_model.eval()
+input_shape = (1, 3, 640, 640)
+flops, macs, params = calculate_flops(
+    model=torch_model, input_shape=input_shape, output_as_string=True, output_precision=4
+)
+
+print("FLOPs:%s    MACs:%s    Params:%s\n" % (flops, macs, params))
```
Comment (saichandax) [inline tests/ttnn/integration_tests/yolox_m/test_yolox_flops.py:17]: flop count calculation is not needed in the PR. keep it locally for reference, and delete this file.

File: models/experimental/functional_yolox_m/demo/ttnn_demo.py [hunk 42]
```diff
@@ -0,0 +1,155 @@
...
+
+    input_tensor3 = torch.permute(patch_bot_right, (0, 2, 3, 1))
+    input_tensor3 = input_tensor3.reshape(
+        input_tensor3.shape[0], 1, input_tensor3.shape[1] * input_tensor3.shape[2], input_tensor3.shape[3]
+    )
+    input_tensor3 = ttnn.from_torch(input_tensor3, dtype=ttnn.bfloat8_b, layout=ttnn.TILE_LAYOUT)
+
+    input_tensor = [input_tensor0, input_tensor1, input_tensor2, input_tensor3]
+    output_tensors = ttnn_model(device, input_tensor)
+    output_tensors = ttnn_model(device, input_tensor)
```
Comment (saichandax) [inline models/experimental/functional_yolox_m/demo/ttnn_demo.py:71]: Why are we doing this twice?

**Comment by punithsekar** [comment]
This PR is up to date.

**Review by arakhmati — APPROVED** [review]

**Review by TT-BrianLiu — APPROVED** [review]

**Review by eyonland — APPROVED** [review]

**Pull request closed without merging**